### ISSUE TITLE
Checkout as a context manager

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#28]
## Expected Behavior
Making checkout as a context manager let the users avoid writing explicit `close()` and we can make sure the lock is released. Keeping the current functionality along with this new ability should be the way to go.

```python
with repo.checkout(write=True) as co:
    co.datasets['dset'] = array
    co.commit()
```
We could either raise a warning if context manager exits without commit or we can even do commit implicitly (by having another argument `autocommit=True` in the `checkout()` call)

## Current Behavior
Currently, we need to explicitly call `close()` for all the write-enabled checkouts.

I could take it up if this doesn't raise any other concerns.

**Comment by rlizzo** [comment]
I like the idea, but this requires some real thought and testing before it could ever be implemented

My main concern is how this would affect `lmdb transaction` handling for the `Dataset` and `Metadata` classes. As noted [In the Documentation](https://hangar-py.readthedocs.io/en/latest/tutorial.html#performance) we get massive performance increases by using context managers for reading/writing data. Without prototyping this, it's not clear to me how entering a CM at the checkout level (on level above datasets or metadata, as shown below) would work:

```
Hangar Object Hierarchy:

Repository
  └ Checkout
      ├ Metadata
      |    ├ MetadataWriter (CM)
      |    └ MetadataReader (CM)    # overridden by MetadataWriter CM if applicable
      └ Dataset
           ├ DatasetDataWriter (CM)
           └ DatasetDataReader (CM)  # overridden by DatasetDataWriter CM if applicable
```

Some of the problems I can see with moving the `CM` enter/exit logic up a level:

1) **Transaction lifespan:** would increase dramatically. The natural implementation would be to reimplement the defined behavior in the checkout object. This would start a single transaction when the `__enter__` method is encountered, which would persist until `__exit__`. Long lived transactions are an antipattern in `lmdb`, as memory page reuse is not allowed while a `txn` is allive. In addition, fatal system crashes would all records added in a checkout (prior to the crash) to be lost, rather than the records related to the dataset/metadata under operation. 
2) **Thread Local Storage:** `lmdb transactions` use `TLS` and can only ever exist in a single thread (the one they were created on). If we want to use multithreaded data loaders in the future, having to keep all record access in a single thread could potentially complicate the logic. 
3) **One Environment Per Process**: An `lmdb environment` can only be opened once in a given process. Since all `transactions` are tied to a single `environment`, supporting multiprocess dataloaders/writers would get complicated (potentially impossible?). This can't even be solved by `closing` and `re-opening` the environment for any spawned processes, since this breaks `flock()` advisory locking on POSIX/UNIX OSs
4) **Storage Backend Support**: This will just be a mess if we attempt to move context management for storage backends up this far. This logic needs to be implemented independently for each backend. We could work around this (in a manner similar to how we do currently), but it would just be another thing which increases complexity of the underslying codebase.

These are just the ones i can come up with off the top of my head. there are probably more...

I think the question we really need to ask is: How important is this? would this feature be worth the pain right now? I personally don't think so, but I'd love to hear your opinion.

**Comment by hhsecond** [comment]
While I really like the user-friendliness of making CM here, you sort of convinced me that moving all the heavy lifting that happens in Dataset/Metadata level CM to repo level is not worth the pain right now. But I think, making repository also a CM while keeping the heavy lifting as it is (i.e. in Dataset/Metadata level) should be helpful. Plus we could keep the normal (naked) way of accessing repository as it is. This might make things weird for some, like calling two `with` statements, etc, but again, having an option doesn't harm. 
```python
with repo.checkout(write=True, autocommit=True) as co:
    with co.datasets['dset'] as dset:
        dset.add(arr, 'arr')
```

**Comment by rlizzo** [comment]
So essentially the entire purpose of the CM checkout block would be to ensure that the writer lock is closed?

**Comment by hhsecond** [comment]
Right (for now) and the user could avoid writing two more lines (for commit and close). Plus, it's up to the user whether they need to use the CM or the naked version.

**Comment by rlizzo** [comment]
Ok. To put this in writing, I'm good with this. I don't know when I'll have time, so I'm going to leave this as help wanted for a bit

**Comment by rlizzo** [comment]
Closed with #84
